File: core/src/main/kotlin/core/ems/service/statistics.kt
package core.ems.service

import com.fasterxml.jackson.annotation.JsonProperty
import core.ems.service.cache.CachingService
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.validation.Valid
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import org.springframework.beans.factory.annotation.Value
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.concurrent.ConcurrentLinkedQueue

private val log = KotlinLogging.logger {}

@RestController
@RequestMapping("/v2")
class StatisticsController(private val statisticsService: StatisticsService) {

    /**
     * Three aggregate counts, readable with no account.
     *
     * Public because the landing page is the first thing a visitor who has never logged in sees, and
     * it has shown these counts all along — what it did not have was permission. The endpoint was
     * `@Secured` for the three roles, so every anonymous visit sat in a five-second retry loop
     * against a 401 for as long as the tab stayed open, writing two lines into core's log per
     * attempt. See the note beside this path in `PERMIT_ALL_PATTERNS` for why publishing three
     * counts is a different decision from publishing component versions.
     *
     * ### Why a POST, and why the request body is the response type
     *
     * This is a long poll, and the body is the caller saying *what it already has*. Matching stats
     * mean there is nothing to send yet, so the request is held until they change — that is how the
     * counters on the page move without the client polling on a timer. A caller with nothing yet
     * sends no body at all and is answered immediately.
     *
     * So the client posts back, verbatim, a body core previously serialised — which makes
     * [StatResp]'s wire names load-bearing in *both* directions. `StatisticsApiTest` posts a
     * received body back for exactly that reason: get the names wrong on one side and every request
     * after the first is a 400, which the client would see only as a poll that never updates.
     */
    @PostMapping("/unauth/statistics/common")
    fun controller(@Valid @RequestBody dto: StatResp?): StatResp {
        log.debug { "Public statistics query" }
        return runBlocking { statisticsService.getOrWaitStatUpdate(dto) }
    }
}

data class StatResp(
    @get:JsonProperty("in_auto_assessing") val inAutoAssessing: Long,
    @get:JsonProperty("total_submissions") val totalSubmissions: Long,
    @get:JsonProperty("total_users") val totalUsers: Long
)


@Service
class StatisticsService(
    private val cachingService: CachingService,
    /**
     * How long a long poll is held before being answered with what we already have.
     *
     * Defaulted rather than added to `application.yaml.sample` and the ansible template: a bound
     * that exists to stop a thread being held forever is not something an operator needs to tune,
     * and `SampleConfigCompletenessTest` deliberately ignores defaulted placeholders so that the
     * sample lists necessities rather than every knob.
     */
    @Value("\${easy.core.statistics.long-poll-timeout.ms:30000}") private val longPollTimeoutMs: Long,
    /**
     * How many callers may be waiting at once. Past this, a caller is answered immediately.
     *
     * The real bound on what long-polling can take from Tomcat's thread pool — see
     * [getOrWaitStatUpdate], where the reasoning is, along with why the timeout is not that bound.
     * Defaulted for the same reason as [longPollTimeoutMs].
     */
    @Value("\${easy.core.statistics.max-waiting-clients:50}") private val maxWaitingClients: Int,
) {
    private val clientsListening = ConcurrentLinkedQueue<Channel<StatResp>>()
    private var lastKnownResponse = composeStatResp()

    /**
     * The caller's stats if they are stale, otherwise a wait — bounded twice over — for an update.
     *
     * ### Waiting occupies a request thread, and the timeout is not what bounds that
     *
     * `runBlocking` here runs on the Tomcat request thread, so a waiting caller holds that thread
     * for the duration. [longPollTimeoutMs] bounds any single request and gives us a moment to
     * notice the caller has gone, but it does **not** bound a client's occupancy: the client
     * re-posts the instant it gets an answer, and a timed-out hold is indistinguishable from an
     * update, so one client sits on close to a whole thread either way.
     *
     * [maxWaitingClients] is the bound that matters. Past it a caller is answered immediately with
     * what we have rather than joining the queue, so long-polling can occupy at most that many of
     * Tomcat's threads however many clients ask — and the rest of the pool stays available to every
     * other endpoint. That is stricter than this code managed while the endpoint was authenticated,
     * when both the queue and the wait were unbounded; it is also the reason publishing the endpoint
     * is not handing anyone a lever. A caller turned away simply polls, which is why the client
     * keeps a floor delay between rounds.
     *
     * ### Leaving the queue
     *
     * The removal is under the same lock [pushStatUpdate] holds, and has to be: that method walks
     * this queue with an emptiness check followed by a `poll()`, and a removal from outside the lock
     * can empty it in between. `ConcurrentLinkedQueue.poll()` is a platform type, so the null would
     * land as an NPE inside a scheduled task rather than anywhere that explains itself.
     *
     * Removing here also closes an older leak: a client that disconnected mid-poll left its channel
     * in the queue forever, because nothing tells a blocked `receive()` that the socket went away.
     *
     * A push landing in the gap between the timeout firing and the removal is not lost work — the
     * value pushed *is* [lastKnownResponse], which is what this then returns.
     */
    suspend fun getOrWaitStatUpdate(currentStatResp: StatResp?): StatResp {
        val channel = Channel<StatResp>(Channel.CONFLATED)

        synchronized(this) {
            // if client has no current stats or stats differ, return latest queried stats...
            if (currentStatResp != lastKnownResponse) return lastKnownResponse
            // ...and likewise if too many are already waiting, rather than taking another thread.
            if (clientsListening.size >= maxWaitingClients) return lastKnownResponse
            clientsListening.add(channel)
        }
        // else wait for update, for as long as we are willing to hold the thread...
        withTimeoutOrNull(longPollTimeoutMs) { channel.receive() }?.let { return it }

        return synchronized(this) {
            clientsListening.remove(channel)
            lastKnownResponse
        }
    }


    @Scheduled(fixedDelayString = "\${easy.core.statistics.fixed-delay.ms}")
    fun pushStatUpdate() {
        synchronized(this) {
            val databaseState = composeStatResp()

            // If nothing new to push, don't push, just return
            if (databaseState == lastKnownResponse) return

            // Has new state, update and push
            lastKnownResponse = databaseState
            // Push response to clients. Null-safe rather than isNotEmpty()/poll(): that pair is
            // atomic only while every removal from the queue is under this lock, which is a property
            // of code elsewhere in this file rather than of this line.
            while (true) {
                val listener = clientsListening.poll() ?: break
                listener.trySend(lastKnownResponse)
            }
        }
    }

    private fun composeStatResp() = StatResp(
        cachingService.countSubmissionsInAutoAssessment(),
        cachingService.countSubmissions(),
        cachingService.countTotalUsers()
    )
}

File: web/src/api/statistics.ts
import { useEffect, useState } from 'react'
import { apiFetch } from './client.ts'

/**
 * The three aggregate counts on the landing page and the About page, on a long poll.
 *
 * ### Public, and called as such
 *
 * `noAuth`, like the other deliberately public reads in `anonymousExercise.ts`: core serves this
 * from `/unauth/statistics/common`, and the landing page is the first thing a visitor who has never
 * logged in sees. It used to call an endpoint that was `@Secured` for the three roles, so an
 * anonymous visitor got a 401, waited five seconds, and asked again — for as long as the tab stayed
 * open, writing two lines into core's log every time round (EZ-1844). Nothing failed; the page just
 * showed a spinner where the counts go, and core's log filled up.
 *
 * ### Long poll, not an interval
 *
 * The request body is what this client already has. Core answers immediately when its counts differ
 * and otherwise holds the request until they change, which is how the counters move without polling
 * on a timer. So the body posted after the first round is a body core itself serialised — the wire
 * names have to work in both directions, and `StatisticsApiTest` posts one back to prove they do.
 *
 * A held request is a held thread on the server, and core bounds how many callers it will hold at
 * once — past that it answers immediately instead. From here the two are indistinguishable, which is
 * exactly why [MIN_ROUND_MS] exists: without a floor, being answered at once would turn this loop
 * into a tight one against an endpoint that now needs no account.
 */

/**
 * @endpoint POST /v2/unauth/statistics/common -> (root)
 * @requestBody POST /v2/unauth/statistics/common
 */
interface Stats {
  in_auto_assessing: number
  total_submissions: number
  total_users: number
}

interface UseStatisticsResult {
  inAutoAssessing: number
  totalSubmissions: number
  totalUsers: number
  isLoading: boolean
  /**
   * The poll gave up without ever getting an answer, so there is no number to show.
   *
   * Separate from `isLoading` because the two want opposite things on screen: a spinner says "any
   * moment now", and after this there is no next attempt. Rendering the `0` these counts fall back
   * to would put a confident falsehood on the front page, which is worse than admitting we do not
   * know.
   *
   * False once anything has arrived, even if the poll later gave up: numbers a few minutes stale are
   * still numbers, and there is nothing useful to tell a visitor about the difference.
   */
  isUnavailable: boolean
}

/**
 * How long to wait before each consecutive retry — and, by running out, when to stop retrying.
 *
 * The loop this feeds had no end at all, which is why a landing page open on a broken endpoint
 * talked to core every five seconds indefinitely. A 401 or a 404 is not a transient error and no
 * number of retries fixes one; a core restart is transient and wants a few. Growing delays serve
 * both: quick enough to ride out a restart, and out of patience within a couple of minutes.
 */
const RETRY_BACKOFF_MS = [2_000, 5_000, 15_000, 30_000, 60_000]

/**
 * The delay before retry number [consecutiveFailures] (1 for the first retry), or `null` to give up.
 *
 * Exported for `tests/unit/statistics-retry.test.mjs`. The hook around it needs a browser and a
 * server to say anything, and the property worth pinning — that this returns `null` rather than
 * going round forever — is a property of a pure function.
 */
export function retryDelayMs(consecutiveFailures: number): number | null {
  if (consecutiveFailures < 1) return null
  return RETRY_BACKOFF_MS[consecutiveFailures - 1] ?? null
}

/**
 * The shortest a successful round may take, so the loop cannot spin.
 *
 * Normally free: core holds the request until the counts change, so a round already takes as long as
 * it takes. It earns its place when core answers at once — a short server-side hold, or a caller
 * turned away because too many are already waiting — where without it the loop would hit a public
 * endpoint as fast as the round trip allows.
 */
const MIN_ROUND_MS = 1_000

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

export function useStatistics(): UseStatisticsResult {
  const [stats, setStats] = useState<Stats | null>(null)
  const [isLoading, setIsLoading] = useState(true)
  const [isUnavailable, setIsUnavailable] = useState(false)
  /** Bumped to start the poll over after it has given up — see the listeners below. */
  const [attempt, setAttempt] = useState(0)

  useEffect(() => {
    let cancelled = false
    let gaveUp = false
    // One controller for the whole loop rather than one per request: it is aborted exactly once,
    // on unmount, and a request in flight at that moment is the only one there can be.
    const controller = new AbortController()

    async function poll() {
      /** What core last told us, i.e. what the next request says it already has. */
      let known: Stats | null = null
      let consecutiveFailures = 0

      while (!cancelled) {
        const startedAt = Date.now()
        try {
          // Annotated, not inferred. `known` appears in this initializer and is assigned from the
          // result below, and tsc reads the pair as a cycle (TS7022) rather than resolving it
          // through the explicit type argument. `undefined` is in the type because apiFetch returns
          // it for a 204 or an empty 200 body, which the next line is about.
          const next: Stats | undefined = await apiFetch<Stats>('/unauth/statistics/common', {
            method: 'POST',
            body: known ?? undefined,
            noAuth: true,
            signal: controller.signal,
          })
          if (cancelled) return

          // An empty body is a failure, not a reading. Storing it would leave `known` empty, so the
          // next request would send no body, so core would answer at once, forever — the tight loop
          // MIN_ROUND_MS also guards, arrived at from the other side.
          if (next == null) throw new Error('empty statistics response')

          setStats(next)
          setIsLoading(false)
          known = next
          consecutiveFailures = 0
        } catch {
          // An abort lands here too, and must not be spent out of the retry budget — the component
          // is going away, so there is nothing left to retry for.
          if (cancelled) return

          consecutiveFailures += 1
          const delay = retryDelayMs(consecutiveFailures)
          if (delay === null) {
            // Out of patience. Both flags matter: `isLoading` false so the spinner stops promising
            // an answer, `isUnavailable` true so the page can say it has none.
            gaveUp = true
            setIsLoading(false)
            setIsUnavailable(true)
            return
          }
          await sleep(delay)
          continue
        }
        await sleep(Math.max(0, MIN_ROUND_MS - (Date.now() - startedAt)))
      }
    }

    poll()

    /**
     * Start over, once, on a signal that something outside may have changed.
     *
     * Re-arming rather than a longer ladder. `core_autodeploy` allows core up to 120 seconds to come
     * back, which outlasts the ladder — so a routine deploy would otherwise leave every open landing
     * page showing a dash until somebody reloaded, and a laptop asleep for three minutes the same.
     * Waking on these events recovers from the transient case while keeping the property the ladder
     * is for: an endpoint that will never answer is asked five times, not forever.
     */
    const rearm = () => {
      if (!gaveUp || document.visibilityState !== 'visible') return
      gaveUp = false
      setIsLoading(true)
      setIsUnavailable(false)
      setAttempt((n) => n + 1)
    }
    window.addEventListener('online', rearm)
    document.addEventListener('visibilitychange', rearm)

    return () => {
      cancelled = true
      controller.abort()
      window.removeEventListener('online', rearm)
      document.removeEventListener('visibilitychange', rearm)
    }
  }, [attempt])

  return {
    inAutoAssessing: stats?.in_auto_assessing ?? 0,
    totalSubmissions: stats?.total_submissions ?? 0,
    totalUsers: stats?.total_users ?? 0,
    // Both gated on having nothing to show, so each means "no number yet, and here is why". Numbers
    // already on screen outrank either: a re-arm after a give-up must not replace real counts with a
    // spinner, and stale counts are better than a dash. `== null` rather than `=== null` because an
    // empty response body reaches this state as undefined.
    isLoading: isLoading && stats == null,
    isUnavailable: isUnavailable && stats == null,
  }
}
